### PR TITLE
[WIP] Implement `BlockManager` backend

### PR DESCRIPTION
**Pull request by seyuboglu**
Overhaul the internals of the Meerkat `DataPanel`. The changes seek to enable:
1. Vectorized row-wise operations (*e.g.* slicing, reduction)
2. Simplified I/O and improved latency
3. Clarified view vs. copy behavior
    - We introduce a [new spec](https://www.notion.so/meerkat-working-doc-40d70d094ac0495684d3fd8ddc809343#2b9460b744a04cbca912c8d017e5887c) detailing when users should expect to get views vs. copies (similar to [this resource](https://numpy.org/doc/stable/reference/arrays.indexing.html) for NumPy) – I'm working on enforcing this spec throughout the codebase. 

The new internals are based primarily off the `BlockManager` class, a dict-like object meant to replace the dictionary we were storing the DataPanel's columns in before. The  `BlockManager` manages links between a DataPanel's columns and data blocks (`AbstractBlock`, `NumpyBlock`) where the data is actually stored. It implements `consolidate`, which takes columns of similar type in a DataPanel and stores their data together in a block, and `apply` which applies row-wise operations (e.g. __getitem__) to the blocks in a vectorized fashion. Other important classes:  
- `BlockRef` objects link a block with the  `BlockManager`. These are critical to the functioning of the BlockManager and are the primary type of object passed between the blocks and the block manager. They consists of two things:
  1. A reference to the block (`self.block`)
  2. A set of columns in the `BlockManager` whose data live in the `Block`
- `BlockableMixin` - a mixin used with `AbstractColumn` that holds references to a column's block and the columns index in the block
- `BlockView` - a simple DataClass holding a block and an index into the block. It is typical for new columns to be created from `BlockView`

Note: I marked this is a WIP because there are still a few more things to be done on this front. 
1. Make `concat` BlockManager aware

Other major changes:
- Removed `visible_rows` from `AbstractColumn`, 
- Removed `_cloneable_kwargs` in favor of a unified `_clone`, `_copy`, and `_view` module (cloneable.py)

### CODE REVIEW

File: meerkat/block/manager.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    def remove(self, name):
+        if name not in self._columns:
+            raise ValueError(f"Remove failed: no column '{name}' in BlockManager.")
+
+        del self._columns[name]
+
+        if name in self._column_to_block_id:
+            # column is blockable
+            block_ref = self._block_refs[self._column_to_block_id[name]]
+            del block_ref[name]
```
Comment (seyuboglu) [inline meerkat/block/manager.py:95]: make this a pop (and everywhere else we do the del) – avoid garbage collecting directly

**Comment by krandiash** [comment]
lgtm!

**Review by krandiash — APPROVED** [review]

**Review by ad12 — APPROVED** [review]
lgtm

**Comment by seyuboglu** [comment]
closes #98

**Pull request merged**